### PR TITLE
Fix #117 

### LINKED ISSUE #117: Provide a central model service component
To make proper use of ChangeBoundsHints and DragAndDropHints (cf. [Protocol Spec](https://github.com/eclipsesource/GraphicalServerProtocol/blob/master/specification.md)) we need a <i>ModelService</i> component that maintains the hints received from the server and can be queried to determine whether a certain operation (move,resize, changeContainer) is allowed for a certain element

### PR DESCRIPTION
**Pull request by tortmayr**
With this change I also propose an update of our protocol definition (cf. https://github.com/eclipsesource/GraphicalServerProtocol/pull/3) which combines the DragAndDropHints and BoundChangeHints as ElementTypeHints. In addition, we can now define via these hints if an element is deletable. 

Based on this protocol changes this PR provides:
+ An inject able service component that maintains all model hints and offers query methods
+ Actions and handlers for requesting hints and populating the service

Note: This PR does not include the integration of the ElementTypeHintService into the different  tools. This has to be done in follow-up PRs

### CODE REVIEW

File: client/packages/glsp-sprotty/src/features/hints/action-definition.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Specifices wheter the element can be deleted
+     */
+    public readonly deletable: boolean;
+
+    /**
+     * The types of elements that can be contained by this element (if any)
+     */
+    public readonly containableElementIds?: string[];
```
Comment (planger) [inline client/packages/glsp-sprotty/src/features/hints/action-definition.ts]: Please rename to `containableElementTypeIds`.

**Comment by tortmayr** [comment]
@planger  I noticed that the client probably needs additional information for edge types (to determine if source and target are connectable)
As some of the existing hint features do not make sense for edges (e.g. resizable, containableElementIds) I propose a separation between NodeTypeHints & EdgeTypeHints. 
What do you think?

**Comment by planger** [comment]
@tortmayr Yes, I think it is a good idea to separate edge and node element type hints. Thanks!

**Comment by tortmayr** [comment]
@planger  I proposed a protocol update PR in https://github.com/eclipsesource/GraphicalServerProtocol/pull/4
Could you check it out. I will make the necessary changes to this once the update PR is merged

**Comment by planger** [comment]
@tortmayr Sure, thanks! I raised one comment in https://github.com/eclipsesource/GraphicalServerProtocol/pull/4. Otherwise looks good. Thanks!

**Comment by tortmayr** [comment]
+ Refactor API and handlers according to new protocol spec
+ Add utils class for arrays
+ Commit-History clean up by squashing into a single commit

File: client/packages/glsp-sprotty/src/features/hints/element-type-hints-service.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+@injectable()
+export class DefaultTypeHintsService implements TypeHintsService {
+    protected typeHints: Map<string, TypeHint> = new Map
+    protected nodeTypeHints: Map<string, NodeTypeHint> = new Map
+    protected edgeTypeHints: Map<string, EdgeTypeHint> = new Map
+    initialize(action: SetTypeHintsAction): void {
+        this.nodeTypeHints = new Map()
+        this.edgeTypeHints = new Map
```
Comment (planger) [inline client/packages/glsp-sprotty/src/features/hints/element-type-hints-service.ts]: For the sake of consistency, it'd be better to be consistent with `new Map()` vs. `new Map`.

File: client/packages/glsp-theia-extension/src/browser/diagram/glsp-theia-diagram-server.ts [hunk 3]
```diff
@@ -24,7 +24,9 @@ export class GLSPTheiaDiagramServer extends TheiaDiagramServer implements Notify
         @inject(TYPES.ViewerOptions) viewerOptions: ViewerOptions,
         @inject(TYPES.SModelStorage) storage: SModelStorage,
         @inject(TYPES.ILogger) logger: ILogger,
-        @inject(ObservableCommandStack) protected commandStack: ObservableCommandStack
+        @inject(ObservableCommandStack) protected commandStack: ObservableCommandStack,
+        @inject(GLSP_TYPES.TypeHintsService) protected typeHintsService: TypeHintsService
+
```
Comment (planger) [inline client/packages/glsp-theia-extension/src/browser/diagram/glsp-theia-diagram-server.ts]: Please remove this empty line.

File: server/glsp-api/src/main/java/com/eclipsesource/glsp/api/action/kind/RequestOperationsAction.java [hunk 4]
```diff
@@ -17,5 +17,7 @@
 	public RequestOperationsAction() {
 		super(Action.Kind.REQUEST_OPERATIONS);
 	}
+	
+
```
Comment (planger) [inline server/glsp-api/src/main/java/com/eclipsesource/glsp/api/action/kind/RequestOperationsAction.java]: Please remove those two added blank lines.

File: server/glsp-api/src/main/java/com/eclipsesource/glsp/api/di/GLSPModule.java [hunk 5]
```diff
@@ -46,14 +46,16 @@ protected final void configure() {
...
-		bind(ModelTypeConfiguration.class).to(bindModelTypeConfiguration());
 		bind(ActionProvider.class).to(bindActionProvider());
 		bind(ActionHandlerProvider.class).to(bindActionHandlerProvider());
 		bind(OperationHandlerProvider.class).to(bindOperatioHandlerProvider());
 		bind(ServerCommandHandlerProvider.class).to(bindServerCommandHandlerProvider());
+		bind(ModelTypeConfigurationProvider.class).to(bindModelTypesConfigurationProvider());
 		configureMultibindings();
 	}
 
+
```
Comment (planger) [inline server/glsp-api/src/main/java/com/eclipsesource/glsp/api/di/GLSPModule.java]: Please remove those two blank lines. We should eventually add some save-actions that auto-format the files including removing more than two consecutive blank lines.

File: client/packages/glsp-sprotty/src/utils/array-utils.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (planger) [inline client/packages/glsp-sprotty/src/utils/array-utils.ts]: I think we should add this class in lib/utils/ to be consistent with the structure in sprotty.

**Comment by tortmayr** [comment]
>I think we should not move the two existing util files back from lib/utils to utils as well as add the new util file in lib/utils in order to be consistent with the structure in sprotty.

In sprotty util files are located in "src/utils" as well (https://github.com/eclipse/sprotty/tree/master/src/utils) so we should already be consistent 😉

**Comment by planger** [comment]
>> I think we should not move the two existing util files back from lib/utils to utils as well as add the new util file in lib/utils in order to be consistent with the structure in sprotty.

> In sprotty util files are located in "src/utils" as well (https://github.com/eclipse/sprotty/tree/master/src/utils) so we should already be consistent wink

Oh, ok... somehow I remembered it otherwise, which is why I moved them into lib originally. Strange... I personally find this distinction between what's in `utils` and what's in `lib` a bit unclear. Aren't `utils` also worth being part of the library?
Anyway, if it is at least consistent to the structure in sprotty, that's fine. I'll resolve the comments regarding this topic above.

**Comment by tortmayr** [comment]
AFAIK in the original sprotty structure the "lib" folder is not a general library but rather refers to their graph abstraction library for SModels. 

<i>(from the sprotty wiki:)</i>
> The base architecture of sprotty does not presume that the visualization is a graph with nodes and edges. As this is a common case, we have added the graph abstractions as a library

So the only model definitions,views etc. that are relevant for this graph abstraction library are located in the "lib" folder.

**Pull request merged**